Added a ViewBoxMixin that lets charts configure their viewbox

samdalton opened the pull request:
Following discussion on this issue here: https://github.com/esbullington/react-d3/issues/204 I implemented a proposed solution and show it working for the AreaChart. The remaining charts have yet to be updated. By using the `ViewBoxMixin` charts can accept a `viewBoxObject` prop that will let charts access outer chart dimensions as pixel values, and generate a string viewBox value that can be set on the svg element. See the AreaChart for an example. Comments/thoughts appreciated - this doesn't need to be merged in but is rather another starting point for the discussions that seemed to have stagnated a bit.

parshap commented on src/mixins/CartesianChartPropsMixin.js:
```diff
@@ -27,12 +27,11 @@ module.exports =  {
-    width:             React.PropTypes.number,
-    height:            React.PropTypes.number,
+    width:             React.PropTypes.node,
+    height:            React.PropTypes.node,
```
I think you want `React.PropTypes.oneOfType([React.PropTypes.number, React.PropTypes.string])` here.

parshap commented on src/common/charts/LegendChart.jsx:
```diff
@@ -70,7 +72,7 @@ module.exports = React.createClass({
         {this._renderTitle()}
         {this._renderLegend()}
-        <svg viewBox={props.viewBox} width={props.width - props.sideOffset} height={props.height}>{props.children}</svg>
+        <svg viewBox={this.getViewBox()} width={this.getOuterDimensions().width - props.sideOffset} height={props.height}>{props.children}</svg>
```
I think `width={this.getOuterDimensions().width}` will result in having a static `px` width set on the svg element instead of a `%` that gets passed in. I see this is a complication because of the arithmetic that needs to be done to account for the legend's size. I think the layout system needs to be changed to use some sort of table layout or floats so that the svg can be `width=100%` while still allowing for room for the legend.

esbullington replied:
There is [this approach](http://blog.scottlogic.com/2015/02/02/svg-layout-flexbox.html) using Facebook's css-layout library to implement/emulate a flexbox layout.  However, this comes with a heavy cost when used with React: the background calculations for SVG could only be done once the component has rendered, which not only will drag down performance, but will likely also cause a visible layout change (a la [FOUC](https://en.wikipedia.org/wiki/Flash_of_unstyled_content)) a few hundred microseconds after the initial page render.

I'm not sure how you could do a table or float layout with all the SVG we use for axes.  SVG unfortunately doesn't support any kind of flow layout.  It might be possible to do a table layout using the SVG `foreignObject` but that's not supported by IE.   It might be possible to pull off a table layout if we limit SVG usage to the chart body only, and have the axes labels as non-svg elements.

Ultimately, complex SVG layouts in React is a hard problem, so reducing the use of SVG to the chart body only might be a wise move (indeed, this is why I have the chart title and key in HTML).

parshap replied:
> I'm not sure how you could do a table or float layout with all the SVG we use for axes. SVG unfortunately doesn't support any kind of flow layout.

We'd only need CSS layout to get the legend (a `div`) to flow alongside of fluid-width `svg` elements.

The overall approach taken in this PR is solid and has achieved fluid-width charts for us (@samdalton and I), the implementation just currently has a bug in `LegendChart`.

yang-wei commented:
Hey @parshap @samdalton, I would really love to merge this but did any of you find a way to implement this in `LegendChart`

samdalton commented:
@yang-wei I've been using this with a custom legend, and a table-layout to contain the chart and legend side by side. I can update LegendChart to use this same technique, but I haven't currently due to not using it.

samdalton commented:
@yang-wei @esbullington I've updated this branch with the `LegendChart` updated to use a table cell layout to scale. I've updated the other relevant legend charts to fully support the ViewBoxMixin and did a bit of clean up on old width calculation logic that is no longer necessary. The docs are updated to show this and it seems to work pretty well!

You'll notice in the example I'm giving the chart a fixed height but a 100% width, this is to keep one dimension consistent, and scale it down on the y axis when resizing.

Please let me know if you have any comments/questions on this approach, but it essentially mirrors what we've got going in production with our fork.

yang-wei commented:
LGTM. Will be merged in #224 since there are some conflict. Thanks again @samdalton.

We might add this into BarChart too when it starts to support multi-series as asked in gitter.
